### PR TITLE
feat: implemented edc data exchange for DemandAndCapacityNotification

### LINKED ISSUE #193: Provide Messaging Interfaces via EDC

### PR DESCRIPTION
**Pull request by ReneSchroederLJ**
## Description
- implemented SAMM model for DemandAndCapacityNotification 2.0
- created the logic for sending Notifications as POST requests through the EDC

resolves #193 and #196 

## Pre-review checks

Please ensure to do as many of the following checks as possible, before asking for committer review:

- [x] DEPENDENCIES are up-to-date. [Dash license tool](https://github.com/eclipse/dash-licenses). Committers can open IP issues for restricted libs.
- [x] [Copyright and license header](https://eclipse-tractusx.github.io/docs/release/trg-7/trg-7-02) are present on all affected files

### CODE REVIEW

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/adapter/DemandAndCapacityNotificationSammMapper.java [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+        var builder = DemandAndCapacityNotificationSamm.builder();
+        var samm = builder
+                .notificationId(notification.getNotificationId().toString())
+                .relatedNotificationId(notification.getRelatedNotificationId() != null ? notification.getRelatedNotificationId().toString() : null)
+                .sourceNotificationId(notification.getSourceNotificationId() != null ? notification.getSourceNotificationId().toString() : null)
+                .text(notification.getText())
+                .leadingRootCause(notification.getLeadingRootCause())
+                .effect(notification.getEffect())
+                .status(notification.getStatus())
+                .materialGlobalAssetId(affectedMaterialGlobalAssetIds)
```
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/adapter/DemandAndCapacityNotificationSammMapper.java:55]: We should add tests to mapping of these edge cases

**Review by tom-rm-meyer-ISST — CHANGES_REQUESTED** [review]
Thanks for contributing! As always great quality. We have some misunderstanding regarding notifications: these are not added as submodels at the registry for a material. 

Let's have a Chat on that :)

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/service/DemandAndCapacityNotifcationRequestApiService.java [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    private ReportedDemandAndCapacityNotificationService reportedDemandAndCapacityNotificationService;
+    @Autowired
+    private EdcAdapterService edcAdapterService;
+    @Autowired
+    private DemandAndCapacityNotificationSammMapper sammMapper;
+    @Autowired
+    private ObjectMapper objectMapper;
+
+    public static final String DEMAND_AND_CAPACITY_NOTIFICATION_CONTEXT = "CX-DemandAndCapacityNotification:2.0";
+    public static final String DEMAND_AND_CAPACITY_NOTIFICATION_VERSION = "3.0.0";
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/service/DemandAndCapacityNotifcationRequestApiService.java:58]: please rename constant to `MESSAGE_HEADER_VERSION`

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/dto/demandandcapacitynotficationsamm/DemandAndCapacityNotificationSamm.java [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+SPDX-License-Identifier: Apache-2.0
+*/
+package org.eclipse.tractusx.puris.backend.demandandcapacitynotification.logic.dto.demandandcapacitynotficationsamm;
+
+import org.eclipse.tractusx.puris.backend.common.util.PatternStore;
+import org.eclipse.tractusx.puris.backend.demandandcapacitynotification.domain.model.EffectEnumeration;
+import org.eclipse.tractusx.puris.backend.demandandcapacitynotification.domain.model.LeadingRootCauseEnumeration;
+import org.eclipse.tractusx.puris.backend.demandandcapacitynotification.domain.model.StatusEnumeration;
+
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/dto/demandandcapacitynotficationsamm/DemandAndCapacityNotificationSamm.java:26]: please remove blank lines in import

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/service/DemandAndCapacityNotifcationRequestApiService.java [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.eclipse.tractusx.puris.backend.demandandcapacitynotification.logic.service;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import lombok.extern.slf4j.Slf4j;
+
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/service/DemandAndCapacityNotifcationRequestApiService.java:26]: please remove blank lines in import

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/adapter/DemandAndCapacityNotificationSammMapper.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/demandandcapacitynotification/logic/adapter/DemandAndCapacityNotificationSammMapper.java:55]: raised issue #465

**Review by tom-rm-meyer-ISST — CHANGES_REQUESTED** [review]
Great refactoring! Only had a few smaller findings. Please merge main; then I'll gonna check it in action :)

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/common/util/VariablesService.java [hunk 5]
```diff
@@ -112,6 +112,20 @@ public class VariablesService {
      */
     private String deliverySubmodelAssetId;
 
+    @Value("${puris.baseurl}" + "catena/demand-and-capacity-notification/request")
```
[the remark this answers is not in the document]
Comment (ReneSchroederLJ) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/common/util/VariablesService.java:115]: catena is part of all asset endpoint variables. If this is an issue we need to address it for all of them I think

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/logic/util/EdcRequestBodyBuilder.java [hunk 6]
```diff
@@ -419,19 +420,39 @@ public JsonNode buildSubmodelRegistrationBody(String assetId, String endpoint, S
...
+    public JsonNode buildNotificationRegistrationBody(String assetId, String endpoint, String semanticId) {
+        var body = getAssetRegistrationContext();
+        body.put("@id", assetId);
+        var propertiesObject = MAPPER.createObjectNode();
+        body.set("properties", propertiesObject);
+        var dctTypeObject = MAPPER.createObjectNode();
+        propertiesObject.set("dct:type", dctTypeObject);
+        dctTypeObject.put("@id", "cx-taxo:DemandAndCapacityNotificationApi");
+        propertiesObject.put("cx-common:version", "1.0");
+        var semanticIdObject = MAPPER.createObjectNode();
```
Comment (tom-rm-meyer-ISST) [inline backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/logic/util/EdcRequestBodyBuilder.java:436]: please remove semanticId object as it's not needed per standard. Also `privateProperties` are not needed.

**Review by tom-rm-meyer-ISST — CHANGES_REQUESTED** [review]
Great work, works as a charm. I just realized that we have the semanticId set in the notificationId Asset Defintion which is not needed.

File: backend/src/main/java/org/eclipse/tractusx/puris/backend/common/util/VariablesService.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (tom-rm-meyer-ISST) [reply backend/src/main/java/org/eclipse/tractusx/puris/backend/common/util/VariablesService.java:115]: True, currently also refactored for existing endpoints here: https://github.com/eclipse-tractusx/puris/pull/443/files#diff-24c2a742f6a2ddaa2cba86e31aa9c04edabba5b10b85d6017e4225f959faa4c1

File: charts/puris/Chart.yaml [hunk 7]
```diff
@@ -35,7 +35,7 @@ dependencies:
 # This is the chart version. This version number should be incremented each time you make changes
 # to the chart and its templates, including the app version.
 # Versions are expected to follow Semantic Versioning (https://semver.org/)
-version: 2.6.4
+version: 2.6.5
```
Comment (tom-rm-meyer-ISST) [inline charts/puris/Chart.yaml:38]: Please change again to 2.6.4. I finally managed to make the job for helm fail but not block merging for not bumping versions. Also updated future PR templates. See #450

**Review by tom-rm-meyer-ISST — CHANGES_REQUESTED** [review]
Thanks a lot, please decrease chart version again, as no 2.6.4 has been released yet

**Review by tom-rm-meyer-ISST — APPROVED** [review]
LGTM, thanks a lot!

Helm version stays as stated in earlier comments
